mmf/models/transformers/backends/huggingface.py

```diff
@@ -0,0 +1,218 @@
+        list_embeddings = []
+        for idx, (token_emb, pos_emb, layer_norm, dropout) in enumerate(
+            zip(
+                self.token_embeddings,
+                self.pos_embeddings,
+                self.layer_norms,
+                self.dropouts,
+            )
+        ):
+            modality_name = self.modality_keys[idx]
+            total_embedding = token_emb(input_ids[modality_name])
+            if modality_name in position_ids:
+                total_embedding += pos_emb(position_ids[modality_name])
+
+            if modality_name in segment_ids:
+                total_embedding += self.token_type_embeddings(
+                    segment_ids[modality_name]
+                )
+
+            list_embeddings.append(dropout(layer_norm(total_embedding)))
+
+        return torch.cat(list_embeddings, dim=1)
+
+
+class HuggingfaceBackend(BaseTransformerBackend):
+    """Transformer backend wih Huggingface transformer models
+    """
+
+    def __init__(self, config: BaseTransformerConfigType, *args, **kwargs):
+        super().__init__(config)
+
+        # Replace transformer layers with scriptable JIT layers
+        replace_with_jit()
+
+    def build_transformer_config(self):
+        self.transformer_config = AutoConfig.from_pretrained(
+            self.config.transformer_base, **OmegaConf.to_container(self.config)
+        )
+
+    def build_transformer_base(self):
+        self.transformer = AutoModel.from_pretrained(
+            self.config.transformer_base, config=self.transformer_config
+        )
+
+    def build_embeddings(self):
+        self.embeddings = HFEmbeddings(
+            self.config, self.transformer_config, self.transformer
+        )
+
+    def get_config(self):
+        return self.transformer_config
+
+    def _transformer_encode(self, embedding, attention_mask):
+        if torch.jit.is_scripting():
+            return self.transformer.encoder(embedding, attention_mask)
+        else:
+            return self.transformer.encoder(
+                embedding, attention_mask, [None] * len(self.transformer.encoder.layer)
+            )
+
+    def forward(
+        self,
+        tokens_ids: Dict[str, Tensor],
+        position_ids: Dict[str, Tensor],
+        segment_ids: Dict[str, Tensor],
+        masks: List[Tensor],
+    ) -> Tuple[Tensor, List[Tensor]]:
+        # Transformer Input Embeddings
+        embedding_output = self.embeddings(
+            input_ids=tokens_ids, position_ids=position_ids, segment_ids=segment_ids
+        )
+
+        # Transformer Attention mask
+        attention_mask = torch.cat(masks, dim=-1)
+        extended_attention_mask = attention_mask.unsqueeze(1).unsqueeze(2)
+        extended_attention_mask = (1.0 - extended_attention_mask) * -10000.0
+
+        # Transformer Encoder
+        encoded_layers = self._transformer_encode(
+            embedding_output,  # combined embedding
+            extended_attention_mask,  # combined attention mask
+        )
```

Comment:
`generate_encoded_layers(embedding_output, attention_mask)` You can pass more stuff in signature to be more generic. But you also can always modified the signature in child class.

All these functions can then go into BaseBackend class